File: testpy/testdurable.py
from durable.lang import *
import redis
import time
import datetime
import traceback
import sys

def unix_time(dt):
    epoch = datetime.datetime.utcfromtimestamp(0)
    delta = dt - epoch
    return int(delta.total_seconds() * 1000.0)

def provide_durability(host, redis_host_name = 'localhost', port = 6379):
    r = redis.StrictRedis(redis_host_name, port, charset="utf-8", decode_responses=True)

    def get_hset_name(ruleset, sid):
        return 'h!{0}!{1}'.format(ruleset, sid)

    def get_list_name(ruleset, sid):
        return 'l!{0}!{1}'.format(ruleset, sid)

    def get_sset_name(ruleset):
        return 's!{0}'.format(ruleset)

    def format_message(action_type, content):
        return '{0},{1}'.format(action_type, content)

    def format_messages(results):
        if not results:
            return '[]'

        messages = ['[']
        for i in range(0, len(results)):
            messages.append(results[i])
            if i < (len(results) - 1):
                messages.append(',')

        messages.append(']') 
        return ''.join(messages)

    def store_message_callback(ruleset, sid, mid, action_type, content):
        try:
            r.hset(get_hset_name(ruleset, sid), mid, format_message(action_type, content))
        except BaseException as e:
            print(e)
            return 601

        return 0

    def delete_message_callback(ruleset, sid, mid):
        try:
            r.hdel(get_hset_name(ruleset, sid), mid)
        except BaseException as e:
            print(e)
            return 602

        return 0

    def queue_message_callback(ruleset, sid, action_type, content):
        try:
            result = r.zscore(get_sset_name(ruleset), sid)
            if not result:
                r.zadd(get_sset_name(ruleset), {sid: unix_time(datetime.datetime.now())})

            r.rpush(get_list_name(ruleset, sid), format_message(action_type, content))
        except BaseException as e:
            print(e)
            return 603 

        return 0

    def get_queued_messages_callback(ruleset, sid):
        try:
            r.zadd(get_sset_name(ruleset), {sid: unix_time(datetime.datetime.now()) + 5000})
            messages = r.lrange(get_list_name(ruleset, sid), 0, -1)
            if len(messages):
                r.delete(get_list_name(ruleset, sid))
                host.complete_get_queued_messages(ruleset, sid, format_messages(messages))
        except BaseException as e:
            print(e)
            return 604

        return 0

    def get_idle_state_callback(ruleset):
        try:
            results = r.zrangebyscore(get_sset_name(ruleset), 0, unix_time(datetime.datetime.now()))
            if (results and len(results)):
                sid = results[0]
                messages = r.hvals(get_hset_name(ruleset, sid))
                host.complete_get_idle_state(ruleset, sid, format_messages(messages))
        except BaseException as e:
            print(e)
            return 606

        return 0
            

    host.set_store_message_callback(store_message_callback)
    host.set_delete_message_callback(delete_message_callback)
    host.set_queue_message_callback(queue_message_callback)
    host.set_get_idle_state_callback(get_idle_state_callback)
    host.set_get_queued_messages_callback(get_queued_messages_callback)


provide_durability(get_host())


with ruleset('test'):
    
    @when_all(c.first << m.subject == 'Hello',
              c.second << m.subject == 'World')
    def say_hello(c):
        if not c.s.count: 
            c.s.count = 1
        else:
            c.s.count += 1

        print('{0} {1} from: {2} count: {3}'.format(c.first.subject, c.second.subject, c.s.sid, c.s.count))

post('test', { 'subject': 'Hello', 'sid': 'a' })
post('test', { 'subject': 'World', 'sid': 'b' })

post('test', { 'subject': 'Hello', 'sid': 'b' })
post('test', { 'subject': 'World', 'sid': 'a' })


time.sleep(30)

File: testjs/testdurable.js
var d = require('../libjs/durable');

var redisLib = require('redis');

var provideDurability = function(host, redisHostName, port, options) {
    var redisClient = redisLib.createClient(port, redisHostName, options);

    var getHsetName = function(ruleset, sid) {
        return 'h!' + ruleset + '!' + sid;
    }

    var getListName = function(ruleset, sid) {
        return 'l!' + ruleset + '!' + sid;
    }

    var getSsetName = function(ruleset) {
        return 's!' + ruleset;
    }

    var formatMessage = function (actionType, content) {
        return actionType + ',' + content;
    }

    var formatMessages = function(results) {
        if (!results) {
            return 0;
        }

        var messages = '[';
        for (var i =0; i < results.length; ++i) {
            messages += results[i];
            if (i < results.length - 1) {
                messages += ',';
            }
        }

        messages += ']';
        return messages;
    }

    var storeMessageCallback = function(ruleset, sid, mid, actionType, content) {
        try {
            redisClient.hset(getHsetName(ruleset,sid), mid, formatMessage(actionType, content));
            return 0;
        } catch (reason) {
            return 601;
        }
    }

    var deleteMessageCallback = function(ruleset, sid, mid) {
        try {
            redisClient.hdel(getHsetName(ruleset,sid), mid);
            return 0;
        } catch (reason) {
            return 602;
        }
    }

    var queueMessageCallback = function(ruleset, sid, actionType, content) {
        try {
            redisClient.zscore(getSsetName(ruleset), sid, function(err, result) {
                if (!err && !result) {
                    var current = new Date() - 0;
                    redisClient.zadd(getSsetName(ruleset), current, sid);
                }
            });
                
            redisClient.rpush(getListName(ruleset,sid), formatMessage(actionType, content));
            return 0;
        } catch (reason) {
            return 603;
        }
    }

    var getQueuedMessagesCallback = function(ruleset, sid) {
        try {
            var current = new Date() - 0;
            redisClient.zadd(getSsetName(ruleset), current + 5000, sid, function(err, result) {
                if (!err) {         
                    redisClient.lrange(getListName(ruleset,sid), 0, -1, function(err, messages) {
                        if (!err && messages.length) {
                            redisClient.del(getListName(ruleset,sid), function(err) {
                                host.completeGetQueuedMessages(ruleset, sid, formatMessages(messages));
                            });
                        }
                    });
                }
            });

            return 0;
        } catch (reason) {
            return 604;
        }
    }

    var getIdleStateCallback = function(ruleset) {
        try {
            var current = new Date() - 0;
            redisClient.zrangebyscore(getSsetName(ruleset), 0, current, function(err, results) {
                if (!err && results.length) {
                    var sid = results[0];
                    redisClient.hvals(getHsetName(ruleset, sid), function(err, messages) {
                        if (!err) {
                            host.completeGetIdleState(ruleset, sid, formatMessages(messages));
                        }
                    });
                }
            });

            return 0;
        } catch (reason) {
            console.log(reason);
            return 606;
        }
    }

    host.setStoreMessageCallback(storeMessageCallback);
    host.setDeleteMessageCallback(deleteMessageCallback);
    host.setQueueMessageCallback(queueMessageCallback);
    host.setGetIdleStateCallback(getIdleStateCallback);
    host.setGetQueuedMessagesCallback(getQueuedMessagesCallback);
}

provideDurability(d.getHost());

d.ruleset('test', function() {
    whenAll: {
        first = m.subject == 'Hello'
        second = m.subject == 'World'
    }
    run: {
        if (!s.count) {
            s.count = 1;
        } else {
            s.count = s.count + 1;
        }

        console.log(first.subject + ' ' + second.subject + ' from: ' + first.sid + ' count: ' + s.count);
    }
});

d.post('test', {subject: 'Hello', sid: 'a'});
d.post('test', {subject: 'World', sid: 'b'});

d.post('test', {subject: 'Hello', sid: 'b'});
d.post('test', {subject: 'World', sid: 'a'});
